core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,263 @@
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.reversedProgression
+ */
+public fun LocalDateProgression.reversed(): LocalDateProgression = LocalDateProgression(longProgression.reversed())
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toLong(), unit)
+
+/**
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
+ * If you wish to switch the direction of the progression, use [LocalDateProgression.reversed]
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(longProgression.step(safeMultiply(value, unit.days.toLong())))
+
+/**
+ * Creates a [LocalDateProgression] from `this` down to [that], inclusive.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.simpleRangeCreation
+ */
+public infix fun LocalDate.downTo(that: LocalDate) : LocalDateProgression = LocalDateProgression.fromClosedRange(this, that, -1, DateTimeUnit.DAY)
+
+/**
+ * Returns a random [LocalDate] within the bounds of the [LocalDateProgression].
+ *
+ * Takes the step into account; will not return any value within the range that would be skipped over by the progression.
+ *
+ * @throws IllegalArgumentException if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.random
+ */
+public fun LocalDateProgression.random(random: Random = Random) : LocalDate = longProgression.random(random).let(LocalDate.Companion::fromEpochDays)
+
+/**
+ * Returns a random [LocalDate] within the bounds of the [LocalDateProgression] or null if the progression is empty.
+ *
+ * Takes the step into account; will not return any value within the range that would be skipped over by the progression.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.random
+ */
+public fun LocalDateProgression.randomOrNull(random: Random = Random) : LocalDate? = longProgression.randomOrNull(random)
+    ?.let(LocalDate.Companion::fromEpochDays)
```

Comment:
These implementations can be made simpler (and more consistent with the other ranges) if they are limited to `LocalDateRange` instead of `LocalDateProgression`. Then, we know that `longProgression` is also a `LongRange`, so we can call `LongRange.random()`.

Reply:
While `longProgression` for a `LocalDateRange` is functionally identical to a `LongRange` it is not, as implemented, technically an instance of that class.

`Range`s are essentially `Progression`s with a step size of `1`, but there is nothing in the constructors or companion creation methods that forces a `Progression` with a step size of `1` to actually _be_ a `Range`. `LongProgression.fromClosedRange(start, end, 1)` will return an instance of `LongProgression`, not a `LongRange` downcast to a `LongProgression`.

We can check if the `step` is `1` and create a `LongRange` if so, but that logic doesn't seem any simpler than what we have now, for reduced functionality.

Reply:
The idea was to cast to a `LongRange`, not to check the `step`; also, with inheritance, we could also specialize `longProgression` to have the `LongRange` type, so the cast wouldn't even be needed.

That said, now that I think about it, since `LocalDateProgression` is a `Collection` now, adding `random` to it seems like a good idea, because otherwise, `random` will still work, but only because `Collection` itself has such an extension function. So, I agree we should keep `LocalDateProgression.random`.